### DISCUSSION TITLE
Is it possible to utilise Node Expansion when returning objects from nodes?

### DISCUSSION THREAD

**Discussion by jonathansandman** [comfyanonymous/ComfyUI#5587] [Q&A]
Hi everybody,

I am utilizing Node Expansion ([docs](https://docs.comfy.org/essentials/custom_node_expansion)). This is a really powerful feature that allows my custom node to return a subgraph of nodes on the fly, taking advantage of the built-in caching functionality in ComfyUI.

However, when returning values from the node:

```
params = {
    "model": merge_model_node.out(0),
    "clip": merge_clip_node.out(0),
    "vae": checkpoint_node1.out(2)
}

return {
    "result": (params, merge_model_node.out(0), merge_clip_node.out(0), checkpoint_node1.out(2)),
    "expand": graph.finalize(),
}
```

If I log the returned value from the node, outputs directly in the tuple appear to be correctly expanded, but those in the object are still the raw links, e.g.:

```
# Example output
(params: {
    "model": ["68.0.0.1", 0],
    "clip": ["72.0.0.1", 0],
    "vae": ["44.0.0.1", 0]
}, Model, CLIP, VAE)
```

where I would expect:

```
(params: {
    "model": Model
    "clip": CLIP,
    "vae": VAE
}, Model, CLIP, VAE)
```

I was hoping that I might be able to utilize Node Expansion and then combine outputs into single object to easily "pipe" it to another node. Is there any way to achieve this?

**Comment by devingfx** [comment]
Hi!
> Disclaimer: I don't know much node expansion, I was searching docs on it when I landed here...

My understanding is that `result` is not working the same as usual return, it just get links from subgraph outputs, and there is an additional routine to 1) execute the subgraph returned ( in `expand` ) and then 2) use the given links to make this node outputs as if it was subgraph outputs.  
So `result` maybe a bit misnamed because you can't give it values, but only "future" links...  
( If I'm not wrong ) maybe renamming this field into something more obvious like `future_links` or so.

--------------

For your case, you can create a small utility node class that takes inputs and outputs a mapping, then use this node in the subgraph, and put its output link in `result`. ( not tested at all ! )
